### PR TITLE
ipc4: fix for copier supported topologies

### PR DESCRIPTION
**Pull request by serhiy-katsyuba-intel**
This fix replaces the previous copier_copy() implementation there
some required input/output combinations were not supported.

TODO: multichannel ALH case when single stream is produced/consumed
by multiple gateways is NOT IMPLEMENTED by this fix. However,
it was not (properly) implemented in the code being replaced.
So the fix does not bring any loss here.

Signed-off-by: Serhiy Katsyuba <serhiy.katsyuba@intel.com>

### CODE REVIEW

File: src/audio/copier/copier.c [hunk 1]
```diff
@@ -775,107 +775,89 @@ static int do_conversion_copy(struct comp_dev *dev,
...
-			cd->output_total_data_processed += processed_data.sink_bytes;
 		}
+	}
 
-		if (!ret) {
-			comp_update_buffer_consume(src_c, processed_data.source_bytes);
-			cd->input_total_data_processed += processed_data.source_bytes;
+	/* zero or more components on outputs */
+	list_for_item(sink_list, &dev->bsink_list) {
+		sink = container_of(sink_list, struct comp_buffer, source_list);
```
Comment (RanderWang) [inline src/audio/copier/copier.c:841]: if gateway is on ouput when component is input, it has been copied in above case. so you need to skip this case in sink_list

Comment (serhiy-katsyuba-intel) [reply src/audio/copier/copier.c:841]: No, actually this case is one of the reasons for the fix. We should support the case of component on input, gateway on output pin 0 with up to 3 more components on outputs 1, 2, 3. This loop is to copy data to these additional components on outputs if present. I described the topologies we have to support in the comment above copier_copy().

File: src/audio/copier/copier.c [hunk 2]
```diff
@@ -775,107 +775,89 @@ static int do_conversion_copy(struct comp_dev *dev,
...
+	}
 
-			if (!ret) {
-				comp_update_buffer_consume(src_c,
-							   processed_data.source_bytes);
-				cd->input_total_data_processed += processed_data.source_bytes;
-			}
+	if (cd->endpoint_num && !cd->bsource_buffer) {
+		/* gateway as input */
+		ret = cd->endpoint[0]->drv->ops.copy(cd->endpoint[0]);
```
Comment (lgirdwood) [inline src/audio/copier/copier.c:810]: We should replace the ```0``` array index with a macro that reflect the default usage, but this can be in another PR.

**Review by lgirdwood — APPROVED** [review]
@RanderWang are you aligned now ?

**Comment by lgirdwood** [comment]
@RanderWang will merge the fix for 2pm daily test inclusion. We can also address any misalignment later if needed.

File: src/audio/copier/copier.c (hunk 2 above)
Comment (RanderWang) [reply src/audio/copier/copier.c:810]: use IPC4_COPIER_GATEWAY_PIN

File: src/audio/copier/copier.c (hunk 1 above)
Comment (RanderWang) [reply src/audio/copier/copier.c:841]: @serhiy-katsyuba-intel , if the gateway is the output, it is not in the sink list, right ? So no problem.

Comment (serhiy-katsyuba-intel) [reply src/audio/copier/copier.c:841]: Yes, gateway buffer is not on sink/source list. However, I do not know if this was specifically designed that way or just happened unintentionally and may change in future. So at least some comment that these buffers in a list are expected not to belong to a gateway might be helpful.
Luckily, this seems is verified on CI: every test that outputs audio would fail as it would produce twice as much data as expected (under a condition we use deep buffers).
